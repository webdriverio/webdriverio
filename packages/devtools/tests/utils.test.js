import {
    validate, getPrototype, findElement, findElements,
    sanitizeError, transformExecuteArgs, transformExecuteResult
} from '../src/utils'

const command = {
    endpoint: '/session/:sessionId/element/:elementId/element',
    method: 'POST',
    command: 'findElementFromElement',
    ref: 'https://w3c.github.io/webdriver/webdriver-spec.html#dfn-find-element-from-element',
    variables: [{
        'name': 'elementId',
        'description': 'the id of an element returned in a previous call to Find Element(s)'
    }],
    parameters: [{
        'name': 'using',
        'type': 'string',
        'description': 'a valid element location strategy',
        'required': true
    }, {
        'name': 'value',
        'type': 'string',
        'description': 'the actual selector that will be used to find an element',
        'required': true
    }, {
        'name': 'customParam',
        'type': 'number',
        'description': 'a random not required param',
        'required': false
    }]
}

let pageMock = {
    waitForSelector: jest.fn(),
    $$eval: jest.fn(),
    $$: jest.fn(),
    $: jest.fn()
}

describe('validate', () => {
    it('should fail if wrong arguments are passed in', () => {
        expect(() => validate(command.command, command.parameters, command.variables, command.refs, []))
            .toThrow(/Wrong parameters applied for findElementFromElement/)
    })

    it('should fail if arguments are malformed', () => {
        expect(() => validate(command.command, command.parameters, command.variables, command.refs, ['123', 123, '123']))
            .toThrow(/Malformed type for "using" parameter of command/)
    })

    it('should fail if not required param has wrong type', () => {
        expect(() => validate(command.command, command.parameters, command.variables, command.refs, ['123', '123', '123', 'foobar']))
            .toThrow(/Malformed type for "customParam" parameter of command/)
    })

    it('should do a proper request', () => {
        const body = validate(command.command, command.parameters, command.variables, command.refs, ['123', 'css selector', '#body', undefined])
        expect(body).toMatchSnapshot()
    })

    it('should do a proper request with non required params', () => {
        const body = validate(command.command, command.parameters, command.variables, command.refs, ['123', 'css selector', '#body', 123])
        expect(body).toMatchSnapshot()
    })
})

test('getPrototype', () => {
    let i = 0
    expect(getPrototype(() => ++i)).toMatchSnapshot()
})

describe('findElement utils', () => {
    afterEach(() => {
        pageMock.waitForSelector.mockClear()
        pageMock.$.mockClear()
        pageMock.$$.mockClear()
    })

    describe('findElement', () => {
        it('tries to find element', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn().mockReturnValue('foobar') }
            }
            pageMock.$.mockReturnValue(Promise.resolve(42))
            expect(await findElement.call(scope, pageMock, 'barfoo'))
                .toEqual({ 'element-6066-11e4-a52e-4f735466cecf': 'foobar' })
            expect(pageMock.waitForSelector).toBeCalledTimes(0)
        })

        it('should fail if not found', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn() }
            }
            pageMock.$.mockReturnValue(Promise.resolve(null))

            const errorMessage = await findElement.call(scope, pageMock, 'barfoo')
            expect(errorMessage.message)
                .toContain('Element with selector "barfoo" not found')
        })

        it('sets implicit waits', async () => {
            const scope = {
                timeouts: { get: jest.fn().mockReturnValue(1234) },
                elementStore: { set: jest.fn() }
            }
            pageMock.$.mockReturnValue(Promise.resolve('foobar'))
            await findElement.call(scope, pageMock, 'barfoo')
            expect(pageMock.waitForSelector).toBeCalledTimes(1)
        })
    })

    describe('findElements', () => {
        it('should find elements', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn().mockReturnValue('foobar') }
            }
            pageMock.$$.mockReturnValue(Promise.resolve([42, 11]))
            expect(await findElements.call(scope, pageMock, 'barfoo')).toMatchSnapshot()
            expect(pageMock.waitForSelector).toBeCalledTimes(0)
        })

        it('should return immiadiatelly if no elements were found', async () => {
            const scope = {
                timeouts: { get: jest.fn() },
                elementStore: { set: jest.fn() }
            }
            pageMock.$$.mockReturnValue(Promise.resolve([]))
            expect(await findElements.call(scope, pageMock, 'barfoo')).toEqual([])
            expect(pageMock.waitForSelector).toBeCalledTimes(0)
            expect(scope.elementStore.set).toBeCalledTimes(0)
        })

        it('sets implicit waits', async () => {
            const scope = {
                timeouts: { get: jest.fn().mockReturnValue(1234) },
                elementStore: { set: jest.fn() }
            }
            pageMock.$$.mockReturnValue(Promise.resolve(['foobar']))
            await findElements.call(scope, pageMock, 'barfoo')
            expect(pageMock.waitForSelector).toBeCalledTimes(1)
        })
    })
})

describe('sanitizeError', () => {
    it('should set stale element error', () => {
        const err = new Error('Node is detached from document bla bla')
        expect(sanitizeError(err).stack)
            .toContain('stale element reference: The element reference is stale')
    })

    it('should not contain puppeteer traces', () => {
        const err = new Error('some error')
        err.stack = err.stack.split('\n')
        err.stack.push('at /foo/bar/devtools/node_modules/puppeteer-core/bla.js')
        err.stack = err.stack.join('\n')
        expect(sanitizeError(err)).not.toContain('devtools/node_modules/puppeteer-core')
    })
})

test('transformExecuteArgs', () => {
    const scope = { elementStore: new Map() }
    scope.elementStore.set('foobar', 'barfoo')

    expect(transformExecuteArgs.call(scope, [
        'foo',
        { 'element-6066-11e4-a52e-4f735466cecf': 'foobar' },
        true,
        42
    ])).toEqual(['foo', 'barfoo', true, 42])
})

describe('transformExecuteResult', () => {
    test('multiple results', async () => {
        const scope = {
            timeouts: { get: jest.fn() },
            elementStore: { set: jest.fn().mockReturnValue('foobar') }
        }
        pageMock.$.mockReturnValue(Promise.resolve(42))
        expect(await transformExecuteResult.call(scope, pageMock, [
            'foobar',
            '__executeElement',
            42
        ])).toMatchSnapshot()
    })

    test('single result', async () => {
        const scope = {
            timeouts: { get: jest.fn() },
            elementStore: { set: jest.fn().mockReturnValue('foobar') }
        }
        pageMock.$.mockReturnValue(Promise.resolve(42))
        expect(await transformExecuteResult.call(scope, pageMock, 'foobar'))
            .toEqual('foobar')
    })

    afterEach(() => {
        pageMock.waitForSelector.mockClear()
        pageMock.$.mockClear()
        pageMock.$$.mockClear()
        pageMock.$$eval.mockClear()
    })
})
