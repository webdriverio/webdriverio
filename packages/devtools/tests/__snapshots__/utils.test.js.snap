// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findElement utils findElements should find elements 1`] = `
Array [
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
]
`;

exports[`findElement utils findElements should find elements with xpath 1`] = `
Array [
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
]
`;

exports[`getPrototype 1`] = `
Object {
  "acceptAlert": Object {
    "value": 52,
  },
  "addCookie": Object {
    "value": 45,
  },
  "back": Object {
    "value": 8,
  },
  "closeWindow": Object {
    "value": 13,
  },
  "createWindow": Object {
    "value": 15,
  },
  "deleteAllCookies": Object {
    "value": 46,
  },
  "deleteCookie": Object {
    "value": 48,
  },
  "deleteSession": Object {
    "value": 2,
  },
  "dismissAlert": Object {
    "value": 51,
  },
  "elementClear": Object {
    "value": 39,
  },
  "elementClick": Object {
    "value": 38,
  },
  "elementSendKeys": Object {
    "value": 40,
  },
  "executeAsyncScript": Object {
    "value": 43,
  },
  "executeScript": Object {
    "value": 42,
  },
  "findElement": Object {
    "value": 24,
  },
  "findElementFromElement": Object {
    "value": 26,
  },
  "findElements": Object {
    "value": 25,
  },
  "findElementsFromElement": Object {
    "value": 27,
  },
  "forward": Object {
    "value": 9,
  },
  "fullscreenWindow": Object {
    "value": 23,
  },
  "getActiveElement": Object {
    "value": 28,
  },
  "getAlertText": Object {
    "value": 53,
  },
  "getAllCookies": Object {
    "value": 44,
  },
  "getElementAttribute": Object {
    "value": 31,
  },
  "getElementCSSValue": Object {
    "value": 33,
  },
  "getElementProperty": Object {
    "value": 32,
  },
  "getElementRect": Object {
    "value": 36,
  },
  "getElementTagName": Object {
    "value": 35,
  },
  "getElementText": Object {
    "value": 34,
  },
  "getNamedCookie": Object {
    "value": 47,
  },
  "getPageSource": Object {
    "value": 41,
  },
  "getTimeouts": Object {
    "value": 4,
  },
  "getTitle": Object {
    "value": 11,
  },
  "getUrl": Object {
    "value": 6,
  },
  "getWindowHandle": Object {
    "value": 12,
  },
  "getWindowHandles": Object {
    "value": 16,
  },
  "getWindowRect": Object {
    "value": 19,
  },
  "isElementDisplayed": Object {
    "value": 30,
  },
  "isElementEnabled": Object {
    "value": 37,
  },
  "isElementSelected": Object {
    "value": 29,
  },
  "maximizeWindow": Object {
    "value": 21,
  },
  "minimizeWindow": Object {
    "value": 22,
  },
  "navigateTo": Object {
    "value": 7,
  },
  "newSession": Object {
    "value": 1,
  },
  "performActions": Object {
    "value": 49,
  },
  "refresh": Object {
    "value": 10,
  },
  "releaseActions": Object {
    "value": 50,
  },
  "sendAlertText": Object {
    "value": 54,
  },
  "setTimeouts": Object {
    "value": 5,
  },
  "setWindowRect": Object {
    "value": 20,
  },
  "status": Object {
    "value": 3,
  },
  "switchToFrame": Object {
    "value": 17,
  },
  "switchToParentFrame": Object {
    "value": 18,
  },
  "switchToWindow": Object {
    "value": 14,
  },
  "takeElementScreenshot": Object {
    "value": 56,
  },
  "takeScreenshot": Object {
    "value": 55,
  },
}
`;

exports[`transformExecuteResult multiple results 1`] = `
Array [
  Array [
    "foobar",
    "__executeElement",
    42,
  ],
  Object {
    "element-6066-11e4-a52e-4f735466cecf": "foobar",
  },
  Array [
    "foobar",
    "__executeElement",
    42,
  ],
]
`;

exports[`validate should do a proper request 1`] = `
Object {
  "elementId": "123",
  "using": "css selector",
  "value": "#body",
}
`;

exports[`validate should do a proper request with non required params 1`] = `
Object {
  "customParam": 123,
  "elementId": "123",
  "using": "css selector",
  "value": "#body",
}
`;
